### ISSUE TITLE
Performance drop in VAE with keras core with respect to tf.keras

### ISSUE THREAD

**Issue by NahuelCostaCortez** [keras-team/keras#19406]
First of all, I would like to thank the Keras team for the amazing work you are doing, seriously.

Having said that, I am migrating most of my autoencoder models to keras 3 and I have encountered the following problems:

- In general, the decay in terms of temporal performance is noticeable (regardless of the backend and model architecture)
- The optimal learning rates change (for example, the same experiment in keras2 with lr=0.001, in keras3 should be 0.01, because 0.001 is not optimal and it gets stuck in a local minima)
- In pytorch, run_eagerly=True has the same performance as run_eagerly=False

Anyhow, the reason I´m opening this issue is mainly because of the first point, which is getting annoying when it comes to VAEs. When I run a VAE in keras 2 the average time per epoch is around 4s whereas if I run exactly the same model in keras3 (the only thing changing being the way the keras Model is defined, which is the first step I´m always doing before changing all operations to be fully back-end agnostic) the average time per epoch goes up to 35s!

Here is a [simple example in Google Colab ](https://gist.github.com/NahuelCostaCortez/c505b2d3fa70d94c7a75dc59ed005444) where you can reproduce the exact behavior. Just make sure the GPU option is selected in Runtime.

Thanks in advance.

**Comment by SuryanarayanaY** [comment]
Hi @NahuelCostaCortez ,

Thanks for reporting. I have replicated the issue with TF2.15(keras2) and TF2.16(Keras3) and observed the same behaviour. The performance seems dropped in Keras3 as per attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/83801120c96d1fc4272190e2f5dd67e5/19406_vae_keras2vskeras3.ipynb).

**Comment by fchollet** [comment]
Thanks for the report. Please bisect the cause of the performance regression by removing features from the code until both versions have the same performance. You can remove the masking layer and the bidirectional wrapper. Then we'll know where the issue is coming from.

**Comment by NahuelCostaCortez** [comment]
Looks like the problem is in the masking layer, see attached [gist](https://colab.research.google.com/gist/NahuelCostaCortez/175037312c7fc547b275d3250ee6515e/19406_vae_keras2vskeras3.ipynb).

**Comment by fchollet** [comment]
I debugging this a bit. The issue isn't exactly with the masking layer, but with the handling of the mask by the LSTM layer.

In Keras 2, we seem to use cuDNN for the LSTM even when the mask was provided. We probably did a value-based check on the mask to verify that it was compatible with cuDNN requirements (the mask needs to only have contiguous 1s on precisely one side and so on). In Keras 3 we're not able to do this check because we need to decided whether to use cuDNN or not before the mask has a defined value.

As a result in Keras 3 we don't use cuDNN when the mask is provided, which results in a performance drop.

Some things we could do:

1. We could a "force_cudnn" argument in LSTM to always attempt to use cuDNN if you know it's supposed to work.
2. You could switch to the JAX backend. I tried it and I get 140ms/step, which isn't quite as good as TF+cuDNN but comes close.
3. You could drop the use of masking, if that works for your use case.

**Comment by NahuelCostaCortez** [comment]
Thank you for the quick feedback. Points 2 and 3 seem reasonable to me. However, I believe the masking layer should work with cuDNN so point 1 looks like a good option. 

My use case is with time series data. I set a predefined sequence length and I mask those series that do not reach that length. Then, I use the masking layer to tell the network not to take these masked values into account. I don´t know if there are any other ways to do this in Keras with LSTMs, but this was a straigthforward approach in Keras 2.

Thanks again.

**Comment by fchollet** [comment]
I have fixed it at HEAD. In addition I added a `use_cudnn` in `LSTM` and `GRU`. It will default to True when it's feasible to use cuDNN.
